lib/vector-core/src/event/finalization.rs

```diff
@@ -147,10 +147,14 @@ impl BatchNotifier {
 
     /// Update this notifier's status from the status of a finalized event.
     fn update_status(&self, status: EventStatus) {
-        // The status starts as Delivered and can only change to Failed
-        // here. A store cycle is much faster than fetch+update.
-        if status == EventStatus::Failed {
-            self.status.store(BatchStatus::Failed, Ordering::Relaxed);
+        // The status starts as Delivered and can only change if the new
+        // status is different than that.
+        if status != EventStatus::Delivered && status != EventStatus::Dropped {
```

Comment:
nitpick: Would it make sense to put this list of variants behind a method that describes what they have in common? That could make this a little easier to scan and maybe help maintain the logic should new variants get added.

Reply:
This optimization is a bit of a cross-domain concern. The set of statuses that can cause of an update is dependent on both the list of event statuses and the initial batch status state. So, the logic belongs in `BatchStatus`, but the only parameter is an `EventStatus`.